File: mysql-async/src/main/scala/com/github/mauricio/async/db/mysql/codec/MySQLConnectionHandler.scala
/*
 * Copyright 2013 Maurício Linhares
 *
 * Maurício Linhares licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

package com.github.mauricio.async.db.mysql.codec

import java.net.InetSocketAddress
import java.nio.ByteBuffer

import com.github.mauricio.async.db.Configuration
import com.github.mauricio.async.db.exceptions.DatabaseException
import com.github.mauricio.async.db.general.MutableResultSet
import com.github.mauricio.async.db.mysql.binary.BinaryRowDecoder
import com.github.mauricio.async.db.mysql.message.client._
import com.github.mauricio.async.db.mysql.message.server._
import com.github.mauricio.async.db.mysql.util.CharsetMapper
import com.github.mauricio.async.db.util.ChannelFutureTransformer.toFuture
import com.github.mauricio.async.db.util._
import io.netty.bootstrap.Bootstrap
import io.netty.buffer.{ByteBuf, ByteBufAllocator, Unpooled}
import io.netty.channel._
import io.netty.channel.socket.nio.NioSocketChannel
import io.netty.handler.codec.CodecException

import scala.annotation.switch
import scala.collection.mutable.{ArrayBuffer, HashMap}
import scala.concurrent._

class MySQLConnectionHandler(
                              configuration: Configuration,
                              charsetMapper: CharsetMapper,
                              handlerDelegate: MySQLHandlerDelegate,
                              group : EventLoopGroup,
                              executionContext : ExecutionContext,
                              connectionId : String
                              )
  extends SimpleChannelInboundHandler[Object] {

  private implicit val internalPool = executionContext
  private final val log = Log.getByName(s"[connection-handler]${connectionId}")
  private final val bootstrap = new Bootstrap().group(this.group)
  private final val connectionPromise = Promise[MySQLConnectionHandler]
  private final val decoder = new MySQLFrameDecoder(configuration.charset, connectionId)
  private final val encoder = new MySQLOneToOneEncoder(configuration.charset, charsetMapper)
  private final val sendLongDataEncoder = new SendLongDataEncoder()
  private final val currentParameters = new ArrayBuffer[ColumnDefinitionMessage]()
  private final val currentColumns = new ArrayBuffer[ColumnDefinitionMessage]()
  private final val parsedStatements = new HashMap[String,PreparedStatementHolder]()
  private final val binaryRowDecoder = new BinaryRowDecoder()

  private var currentPreparedStatementHolder : PreparedStatementHolder = null
  private var currentPreparedStatement : PreparedStatement = null
  private var currentQuery : MutableResultSet[ColumnDefinitionMessage] = null
  private var currentContext: ChannelHandlerContext = null

  def connect: Future[MySQLConnectionHandler] = {
    this.bootstrap.channel(classOf[NioSocketChannel])
    this.bootstrap.handler(new ChannelInitializer[io.netty.channel.Channel]() {

      override def initChannel(channel: io.netty.channel.Channel): Unit = {
        channel.pipeline.addLast(
          decoder,
          encoder,
          sendLongDataEncoder,
          MySQLConnectionHandler.this)
      }

    })

    this.bootstrap.option[java.lang.Boolean](ChannelOption.SO_KEEPALIVE, true)
    this.bootstrap.option[ByteBufAllocator](ChannelOption.ALLOCATOR, LittleEndianByteBufAllocator.INSTANCE)

    this.bootstrap.connect(new InetSocketAddress(configuration.host, configuration.port)).onFailure {
      case exception => this.connectionPromise.tryFailure(exception)
    }

    this.connectionPromise.future
  }

  override def channelRead0(ctx: ChannelHandlerContext, message: Object) {
    message match {
      case m: ServerMessage => {
        (m.kind: @switch) match {
          case ServerMessage.ServerProtocolVersion => {
            handlerDelegate.onHandshake(m.asInstanceOf[HandshakeMessage])
          }
          case ServerMessage.Ok => {
            this.clearQueryState
            handlerDelegate.onOk(m.asInstanceOf[OkMessage])
          }
          case ServerMessage.Error => {
            this.clearQueryState
            handlerDelegate.onError(m.asInstanceOf[ErrorMessage])
          }
          case ServerMessage.EOF => {
            this.handleEOF(m)
          }
          case ServerMessage.ColumnDefinition => {
            val message = m.asInstanceOf[ColumnDefinitionMessage]

            if ( currentPreparedStatementHolder != null && this.currentPreparedStatementHolder.needsAny ) {
              this.currentPreparedStatementHolder.add(message)
            }

            this.currentColumns += message
          }
          case ServerMessage.ColumnDefinitionFinished => {
            this.onColumnDefinitionFinished()
          }
          case ServerMessage.PreparedStatementPrepareResponse => {
            this.onPreparedStatementPrepareResponse(m.asInstanceOf[PreparedStatementPrepareResponse])
          }
          case ServerMessage.Row => {
            val message = m.asInstanceOf[ResultSetRowMessage]
            val items = new Array[Any](message.size)

            var x = 0
            while ( x < message.size ) {
              items(x) = if ( message(x) == null ) {
                null
              } else {
                val columnDescription = this.currentQuery.columnTypes(x)
                columnDescription.textDecoder.decode(columnDescription, message(x), configuration.charset)
              }
              x += 1
            }

            this.currentQuery.addRow(items)
          }
          case ServerMessage.BinaryRow => {
            val message = m.asInstanceOf[BinaryRowMessage]
            this.currentQuery.addRow( this.binaryRowDecoder.decode(message.buffer, this.currentColumns ))
          }
          case ServerMessage.ParamProcessingFinished => {
          }
          case ServerMessage.ParamAndColumnProcessingFinished => {
            this.onColumnDefinitionFinished()
          }
        }
      }
    }

  }

  override def channelActive(ctx: ChannelHandlerContext): Unit = {
    log.debug("Channel became active")
    handlerDelegate.connected(ctx)
  }


  override def channelInactive(ctx: ChannelHandlerContext) {
    log.debug("Channel became inactive")
  }

  override def exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable) {
    // unwrap CodecException if needed
    cause match {
      case t: CodecException => handleException(t.getCause)
      case _ =>  handleException(cause)
    }

  }

  private def handleException(cause: Throwable) {
    if (!this.connectionPromise.isCompleted) {
      this.connectionPromise.failure(cause)
    }
    handlerDelegate.exceptionCaught(cause)
  }

  override def handlerAdded(ctx: ChannelHandlerContext) {
    this.currentContext = ctx
  }

  def write( message : QueryMessage ) : ChannelFuture = {
    this.decoder.queryProcessStarted()
    writeAndHandleError(message)
  }

  def sendPreparedStatement( query: String, values: Seq[Any] ): Future[ChannelFuture] = {
    val preparedStatement = new PreparedStatement(query, values)

    this.currentColumns.clear()
    this.currentParameters.clear()

    this.currentPreparedStatement = preparedStatement

    this.parsedStatements.get(preparedStatement.statement) match {
      case Some( item ) => {
        this.executePreparedStatement(item.statementId, item.columns.size, preparedStatement.values, item.parameters)
      }
      case None => {
        decoder.preparedStatementPrepareStarted()
        writeAndHandleError( new PreparedStatementPrepareMessage(preparedStatement.statement) )
      }
    }
  }

  def write( message : HandshakeResponseMessage ) : ChannelFuture = {
    decoder.hasDoneHandshake = true
    writeAndHandleError(message)
  }

  def write( message : AuthenticationSwitchResponse ) : ChannelFuture = writeAndHandleError(message)

  def write( message : QuitMessage ) : ChannelFuture = {
    writeAndHandleError(message)
  }

  def disconnect: ChannelFuture = this.currentContext.close()

  def clearQueryState {
    this.currentColumns.clear()
    this.currentParameters.clear()
    this.currentQuery = null
  }

  def isConnected : Boolean = {
    if ( this.currentContext != null && this.currentContext.channel() != null ) {
      this.currentContext.channel.isActive
    } else {
      false
    }
  }

  private def executePreparedStatement( statementId : Array[Byte], columnsCount : Int, values : Seq[Any], parameters : Seq[ColumnDefinitionMessage] ): Future[ChannelFuture] = {
    decoder.preparedStatementExecuteStarted(columnsCount, parameters.size)
    this.currentColumns.clear()
    this.currentParameters.clear()

    val (nonLongIndicesOpt, longValuesOpt) = values.zipWithIndex.map {
      case (Some(value), index) if isLong(value) => (None, Some(index, value))
      case (value, index) if isLong(value) => (None, Some(index, value))
      case (_, index) => (Some(index), None)
    }.unzip
    val nonLongIndices: Seq[Int] = nonLongIndicesOpt.flatten
    val longValues: Seq[(Int, Any)] = longValuesOpt.flatten

    if (longValues.nonEmpty) {
      val (firstIndex, firstValue) = longValues.head
      var channelFuture: Future[ChannelFuture] = sendLongParameter(statementId, firstIndex, firstValue)
      longValues.tail foreach { case (index, value) =>
        channelFuture = channelFuture.flatMap { _ =>
          sendLongParameter(statementId, index, value)
        }
      }
      channelFuture flatMap { _ =>
        writeAndHandleError(new PreparedStatementExecuteMessage(statementId, values, nonLongIndices.toSet, parameters))
      }
    } else {
      writeAndHandleError(new PreparedStatementExecuteMessage(statementId, values, nonLongIndices.toSet, parameters))
    }
  }

  private def isLong(value: Any): Boolean = {
    value match {
      case v : Array[Byte] => v.length > SendLongDataEncoder.LONG_THRESHOLD
      case v : ByteBuffer => v.remaining() > SendLongDataEncoder.LONG_THRESHOLD
      case v : ByteBuf => v.readableBytes() > SendLongDataEncoder.LONG_THRESHOLD

      case _ => false
    }
  }

  private def sendLongParameter(statementId: Array[Byte], index: Int, longValue: Any): Future[ChannelFuture] = {
    longValue match {
      case v : Array[Byte] =>
        sendBuffer(Unpooled.wrappedBuffer(v), statementId, index)

      case v : ByteBuffer =>
        sendBuffer(Unpooled.wrappedBuffer(v), statementId, index)

      case v : ByteBuf =>
        sendBuffer(v, statementId, index)
    }
  }

  private def sendBuffer(buffer: ByteBuf, statementId: Array[Byte], paramId: Int): ChannelFuture = {
    writeAndHandleError(new SendLongDataMessage(statementId, buffer, paramId))
  }

  private def onPreparedStatementPrepareResponse( message : PreparedStatementPrepareResponse ) {
    this.currentPreparedStatementHolder = new PreparedStatementHolder( this.currentPreparedStatement.statement, message)
  }

  def onColumnDefinitionFinished() {

    val columns = if ( this.currentPreparedStatementHolder != null ) {
      this.currentPreparedStatementHolder.columns
    } else {
      this.currentColumns
    }

    this.currentQuery = new MutableResultSet[ColumnDefinitionMessage](columns)

    if ( this.currentPreparedStatementHolder != null ) {
      this.parsedStatements.put( this.currentPreparedStatementHolder.statement, this.currentPreparedStatementHolder )
      this.executePreparedStatement(
        this.currentPreparedStatementHolder.statementId,
        this.currentPreparedStatementHolder.columns.size,
        this.currentPreparedStatement.values,
        this.currentPreparedStatementHolder.parameters
      )
      this.currentPreparedStatementHolder = null
      this.currentPreparedStatement = null
    }
  }

  private def writeAndHandleError( message : Any ) : ChannelFuture =  {

    if ( this.currentContext.channel().isActive ) {
      val future = this.currentContext.writeAndFlush(message)

      future.onFailure {
        case e : Throwable => handleException(e)
      }

      future
    } else {
      throw new DatabaseException("This channel is not active and can't take messages")
    }
  }

  private def handleEOF( m : ServerMessage ) {
    m match {
      case eof : EOFMessage => {
        val resultSet = this.currentQuery
        this.clearQueryState

        if ( resultSet != null ) {
          handlerDelegate.onResultSet( resultSet, eof )
        } else {
          handlerDelegate.onEOF(eof)
        }
      }
      case authenticationSwitch : AuthenticationSwitchRequest => {
        handlerDelegate.switchAuthentication(authenticationSwitch)
      }
    }
  }

}

File: mysql-async/src/main/java/com/github/jasync/sql/db/mysql/codec/MySQLConnectionHandler.kt
package com.github.jasync.sql.db.mysql.codec

import com.github.jasync.sql.db.Configuration
import com.github.jasync.sql.db.exceptions.DatabaseException
import com.github.jasync.sql.db.general.MutableResultSet
import com.github.jasync.sql.db.mysql.binary.BinaryRowDecoder
import com.github.jasync.sql.db.mysql.encoder.auth.AuthenticationMethod
import com.github.jasync.sql.db.mysql.message.client.AuthenticationSwitchResponse
import com.github.jasync.sql.db.mysql.message.client.CapabilityRequestMessage
import com.github.jasync.sql.db.mysql.message.client.CloseStatementMessage
import com.github.jasync.sql.db.mysql.message.client.HandshakeResponseMessage
import com.github.jasync.sql.db.mysql.message.client.PreparedStatementExecuteMessage
import com.github.jasync.sql.db.mysql.message.client.PreparedStatementPrepareMessage
import com.github.jasync.sql.db.mysql.message.client.QueryMessage
import com.github.jasync.sql.db.mysql.message.client.QuitMessage
import com.github.jasync.sql.db.mysql.message.client.SendLongDataMessage
import com.github.jasync.sql.db.mysql.message.server.AuthMoreDataMessage
import com.github.jasync.sql.db.mysql.message.server.AuthenticationSwitchRequest
import com.github.jasync.sql.db.mysql.message.server.BinaryRowMessage
import com.github.jasync.sql.db.mysql.message.server.ColumnDefinitionMessage
import com.github.jasync.sql.db.mysql.message.server.EOFMessage
import com.github.jasync.sql.db.mysql.message.server.ErrorMessage
import com.github.jasync.sql.db.mysql.message.server.HandshakeMessage
import com.github.jasync.sql.db.mysql.message.server.OkMessage
import com.github.jasync.sql.db.mysql.message.server.PreparedStatementPrepareResponse
import com.github.jasync.sql.db.mysql.message.server.ResultSetRowMessage
import com.github.jasync.sql.db.mysql.message.server.ServerMessage
import com.github.jasync.sql.db.mysql.util.CapabilityFlag
import com.github.jasync.sql.db.mysql.util.CharsetMapper
import com.github.jasync.sql.db.util.ExecutorServiceUtils
import com.github.jasync.sql.db.util.FP
import com.github.jasync.sql.db.util.NettyUtils
import com.github.jasync.sql.db.util.XXX
import com.github.jasync.sql.db.util.failed
import com.github.jasync.sql.db.util.flatMapAsync
import com.github.jasync.sql.db.util.head
import com.github.jasync.sql.db.util.installOnFuture
import com.github.jasync.sql.db.util.length
import com.github.jasync.sql.db.util.onFailure
import com.github.jasync.sql.db.util.tail
import com.github.jasync.sql.db.util.toCompletableFuture
import io.netty.bootstrap.Bootstrap
import io.netty.buffer.ByteBuf
import io.netty.buffer.Unpooled
import io.netty.channel.Channel
import io.netty.channel.ChannelFuture
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.ChannelInitializer
import io.netty.channel.ChannelOption
import io.netty.channel.EventLoopGroup
import io.netty.channel.SimpleChannelInboundHandler
import io.netty.handler.codec.CodecException
import mu.KotlinLogging
import java.nio.ByteBuffer
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Executor

private val logger = KotlinLogging.logger {}

class MySQLConnectionHandler(
    val configuration: Configuration,
    charsetMapper: CharsetMapper,
    private val handlerDelegate: MySQLHandlerDelegate,
    private val group: EventLoopGroup,
    private val executionContext: Executor = ExecutorServiceUtils.CommonPool,
    private val connectionId: String
) : SimpleChannelInboundHandler<Any>() {

    private val bootstrap = Bootstrap().group(this.group)
    private val connectionPromise = CompletableFuture<MySQLConnectionHandler>()
    private val decoder = MySQLFrameDecoder(configuration.charset, connectionId)
    private val encoder = MySQLOneToOneEncoder(configuration.charset, charsetMapper)
    private val sendLongDataEncoder = SendLongDataEncoder()
    private val currentColumns = mutableListOf<ColumnDefinitionMessage>()
    private val parsedStatements = HashMap<String, PreparedStatementHolder>()
    private val binaryRowDecoder = BinaryRowDecoder()

    private var sslEstablished: Boolean = false
    private var currentPreparedStatementHolder: PreparedStatementHolder? = null
    private var currentPreparedStatement: PreparedStatement? = null
    private var currentQuery: MutableResultSet<ColumnDefinitionMessage>? = null
    private var currentContext: ChannelHandlerContext? = null
    private var currentQueryString: String? = null
    private var isPreparedStatement: Boolean? = null

    fun connect(): CompletableFuture<MySQLConnectionHandler> {
        val socketChannelAddress = NettyUtils.getSocketChannelClassAndSocketAddress(
            this.group,
            configuration
        )
        this.bootstrap.channel(socketChannelAddress.socketChannelClass)
        this.bootstrap.handler(object : ChannelInitializer<Channel>() {

            override fun initChannel(channel: Channel) {
                channel.pipeline().addLast(
                    decoder,
                    encoder,
                    sendLongDataEncoder,
                    this@MySQLConnectionHandler
                )
            }
        })

        this.bootstrap.option(ChannelOption.SO_KEEPALIVE, true)
        this.bootstrap.option(ChannelOption.ALLOCATOR, LittleEndianByteBufAllocator.INSTANCE)
        this.bootstrap.option(ChannelOption.CONNECT_TIMEOUT_MILLIS, this.configuration.connectionTimeout)

        val channelFuture: ChannelFuture = this.bootstrap.connect(socketChannelAddress.socketAddress)
        channelFuture.onFailure(executionContext) { exception ->
            this.connectionPromise.completeExceptionally(exception)
        }

        return this.connectionPromise
    }

    override fun channelRead0(ctx: ChannelHandlerContext, message: Any) {
        when (message) {
            is ServerMessage -> {
                when (message.kind) {
                    ServerMessage.ServerProtocolVersion -> {
                        handlerDelegate.onHandshake(message as HandshakeMessage)
                    }
                    ServerMessage.Ok -> {
                        this.clearQueryState()
                        handlerDelegate.onOk(message as OkMessage)
                    }
                    ServerMessage.Error -> {
                        this.clearQueryState()
                        handlerDelegate.onError(message as ErrorMessage)
                    }
                    ServerMessage.EOF -> {
                        this.handleEOF(message)
                    }
                    ServerMessage.AuthMoreData -> {
                        val m = message as AuthMoreDataMessage

                        if (!m.isSuccess()) {
                            if (!sslEstablished) {
                                throw IllegalStateException(
                                    "Full authentication mode for ${AuthenticationMethod.CachingSha2} requires SSL"
                                )
                            }

                            val request = AuthenticationSwitchRequest(AuthenticationMethod.CachingSha2, null)
                            handlerDelegate.switchAuthentication(request)
                        }
                    }
                    ServerMessage.ColumnDefinition -> {
                        val m = message as ColumnDefinitionMessage

                        this.currentPreparedStatementHolder?.let {
                            if (it.needsAny()) {
                                it.add(m)
                            }
                        }

                        this.currentColumns += message
                    }
                    ServerMessage.ColumnDefinitionFinished -> {
                        this.onColumnDefinitionFinished()
                    }
                    ServerMessage.PreparedStatementPrepareResponse -> {
                        this.onPreparedStatementPrepareResponse(message as PreparedStatementPrepareResponse)
                    }
                    ServerMessage.Row -> {
                        val rsrMessage = message as ResultSetRowMessage
                        val items = Array(rsrMessage.size) {
                            if (rsrMessage[it] == null) {
                                null
                            } else {
                                if (this.currentQuery == null) {
                                    throw NullPointerException("currentQuery is null")
                                }
                                val columnDescription: ColumnDefinitionMessage = this.currentQuery!!.columnTypes[it]
                                val buf = rsrMessage[it]!!
                                try {
                                    columnDescription.textDecoder.decode(
                                        columnDescription,
                                        buf,
                                        configuration.charset
                                    )
                                } finally {
                                    buf.release()
                                }
                            }
                        }

                        this.currentQuery!!.addRow(items)
                    }
                    ServerMessage.BinaryRow -> {
                        val m = message as BinaryRowMessage
                        try {
                            val decodedRow = this.binaryRowDecoder.decode(m.buffer, this.currentColumns)
                            this.currentQuery!!.addRow(decodedRow)
                        } finally {
                            m.buffer.release()
                        }
                    }
                    ServerMessage.ParamProcessingFinished -> {
                    }
                    ServerMessage.ParamAndColumnProcessingFinished -> {
                        this.onColumnDefinitionFinished()
                    }
                }
            }
        }
    }

    override fun channelActive(ctx: ChannelHandlerContext) {
        logger.trace { "[connectionId:$connectionId] - Channel became active" }
        handlerDelegate.connected(ctx)
    }

    override fun channelInactive(ctx: ChannelHandlerContext) {
        logger.trace { "[connectionId:$connectionId] - Channel became inactive" }
    }

    override fun channelRegistered(ctx: ChannelHandlerContext) {
        logger.trace { "[connectionId:$connectionId] - channelRegistered" }
        super.channelRegistered(ctx)
    }

    override fun channelUnregistered(ctx: ChannelHandlerContext) {
        logger.trace { "[connectionId:$connectionId] - channelUnregistered" }
        handlerDelegate.unregistered()
        super.channelUnregistered(ctx)
    }

    @Suppress("OverridingDeprecatedMember")
    override fun exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable) {
        // unwrap CodecException if needed
        when (cause) {
            is CodecException -> handleException(cause.cause ?: cause)
            else -> handleException(cause)
        }
    }

    private fun handleException(cause: Throwable) {
        if (!this.connectionPromise.isDone) {
            this.connectionPromise.failed(cause)
        }
        handlerDelegate.exceptionCaught(cause)
    }

    override fun handlerAdded(ctx: ChannelHandlerContext) {
        this.currentContext = ctx
    }

    private fun write(message: QueryMessage): ChannelFuture {
        this.decoder.queryProcessStarted()
        return writeAndHandleError(message)
    }

    fun sendQuery(query: String) {
        this.isPreparedStatement = false
        this.currentQueryString = query
        this.write(QueryMessage(query))
    }

    fun sendPreparedStatement(query: String, values: List<Any?>): CompletableFuture<ChannelFuture> {
        val preparedStatement = PreparedStatement(query, values)
        this.isPreparedStatement = true
        this.currentQueryString = query
        this.currentColumns.clear()

        this.currentPreparedStatement = preparedStatement

        val item = this.parsedStatements[preparedStatement.statement]
        return when {
            item != null -> {
                this.executePreparedStatement(
                    item.statementId(),
                    item.columns.size,
                    preparedStatement.values,
                    item.parameters
                )
            }
            else -> {
                decoder.preparedStatementPrepareStarted()
                writeAndHandleError(PreparedStatementPrepareMessage(preparedStatement.statement)).toCompletableFuture()
            }
        }
    }

    fun closePreparedStatement(query: String): CompletableFuture<Boolean> {
        val statement = this.parsedStatements[query]
        return if (statement != null) {
            this.parsedStatements.remove(query)
            this.writeAndHandleError(CloseStatementMessage(statement.statementId()))
            FP.successful(true)
        } else {
            FP.successful(false)
        }
    }

    fun write(message: CapabilityRequestMessage): ChannelFuture = writeAndHandleError(message)

    fun write(message: HandshakeResponseMessage): ChannelFuture {
        sslEstablished = message.header.flags.contains(CapabilityFlag.CLIENT_SSL)
        decoder.hasDoneHandshake = true
        return writeAndHandleError(message)
    }

    fun write(message: AuthenticationSwitchResponse): ChannelFuture = writeAndHandleError(message)

    fun sendQuitMessage(): CompletableFuture<Channel> {
        val future = CompletableFuture<Channel>()
        val channel = this.currentContext!!.channel()
        channel.eventLoop().execute {
            this.clearQueryState()
            if (channel.isActive) {
                writeAndHandleError(QuitMessage.Instance).installOnFuture(future)
            } else {
                future.complete(channel)
            }
        }
        return future
    }

    fun closeChannel(): ChannelFuture {
        return this.currentContext!!.channel().close()
    }

    private fun clearQueryState() {
        this.currentColumns.clear()
        this.currentQuery = null
        this.isPreparedStatement = null
        this.currentQueryString = null
    }

    fun isConnected(): Boolean {
        return this.currentContext?.channel()?.isActive ?: false
    }

    private fun executePreparedStatement(
        statementId: ByteArray,
        columnsCount: Int,
        values: List<Any?>,
        parameters: List<ColumnDefinitionMessage>
    ): CompletableFuture<ChannelFuture> {
        decoder.preparedStatementExecuteStarted(columnsCount, parameters.size)
        this.currentColumns.clear()
        val (longValues1, nonLongIndicesOpt1) = values.mapIndexed { index, any -> index to any }
            .partition { (_, any) -> any != null && isLong(any) }
        val nonLongIndices: List<Int> = nonLongIndicesOpt1.map { it.first }
        val longValues: List<Pair<Int, Any>> =
            longValues1.mapNotNull { if (it.second == null) null else it.first to it.second!! }

        return if (longValues.isNotEmpty()) {
            val (firstIndex, firstValue) = longValues.head
            var channelFuture: CompletableFuture<ChannelFuture> = sendLongParameter(statementId, firstIndex, firstValue)
            longValues.tail.forEach { (index, value) ->
                channelFuture = channelFuture.flatMapAsync(executionContext) {
                    sendLongParameter(statementId, index, value)
                }
            }
            channelFuture.toCompletableFuture().flatMapAsync(executionContext) {
                writeAndHandleError(
                    PreparedStatementExecuteMessage(
                        statementId,
                        values,
                        nonLongIndices.toSet(),
                        parameters
                    )
                ).toCompletableFuture()
            }
        } else {
            writeAndHandleError(
                PreparedStatementExecuteMessage(
                    statementId,
                    values,
                    nonLongIndices.toSet(),
                    parameters
                )
            ).toCompletableFuture()
        }
    }

    private fun isLong(value: Any): Boolean {
        return when (value) {
            is ByteArray -> value.length > SendLongDataEncoder.LONG_THRESHOLD
            is ByteBuffer -> value.remaining() > SendLongDataEncoder.LONG_THRESHOLD
            is ByteBuf -> value.readableBytes() > SendLongDataEncoder.LONG_THRESHOLD
            else -> false
        }
    }

    private fun sendLongParameter(
        statementId: ByteArray,
        index: Int,
        longValue: Any
    ): CompletableFuture<ChannelFuture> {
        return when (longValue) {
            is ByteArray ->
                sendBuffer(Unpooled.wrappedBuffer(longValue), statementId, index)

            is ByteBuffer ->
                sendBuffer(Unpooled.wrappedBuffer(longValue), statementId, index)

            is ByteBuf ->
                sendBuffer(longValue, statementId, index)
            else -> XXX("no handle for ${longValue::class.java}")
        }.toCompletableFuture()
    }

    private fun sendBuffer(buffer: ByteBuf, statementId: ByteArray, paramId: Int): ChannelFuture {
        return writeAndHandleError(SendLongDataMessage(statementId, buffer, paramId))
    }

    private fun onPreparedStatementPrepareResponse(message: PreparedStatementPrepareResponse) {
        this.currentPreparedStatementHolder =
            PreparedStatementHolder(this.currentPreparedStatement!!.statement, message)
    }

    private fun onColumnDefinitionFinished() {
        logger.trace { "[connectionId:$connectionId] - onColumnDefinitionFinished()" }

        val columns =
            this.currentPreparedStatementHolder?.columns ?: this.currentColumns

        this.currentQuery = MutableResultSet(columns.toList())

        this.currentPreparedStatementHolder?.let {
            this.parsedStatements[it.statement] = it
            this.executePreparedStatement(
                it.statementId(),
                it.columns.size,
                this.currentPreparedStatement!!.values,
                it.parameters
            )
            this.currentPreparedStatementHolder = null
            this.currentPreparedStatement = null
        }
    }

    private fun writeAndHandleError(message: Any): ChannelFuture {
        return if (currentContext?.channel()?.isActive == true) {
            val res: ChannelFuture = currentContext!!.writeAndFlush(message)

            res.onFailure(executionContext) { e: Throwable ->
                handleException(e)
            }

            res
        } else {
            val error = DatabaseException("This channel is not active and can't take messages")
            handleException(error)
            currentContext!!.channel().newFailedFuture(error)
        }
    }

    private fun handleEOF(m: ServerMessage) {
        when (m) {
            is EOFMessage -> {
                val resultSet = this.currentQuery
                this.clearQueryState()

                if (resultSet != null) {
                    handlerDelegate.onResultSet(resultSet, m)
                } else {
                    handlerDelegate.onEOF(m)
                }
            }
            is AuthenticationSwitchRequest -> {
                handlerDelegate.switchAuthentication(m)
            }
        }
    }
}
